### PR TITLE
Icarl reproducible cl

### PR DESCRIPTION
**Pull request by rudysemola**
Add iCaRL experiment

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -44,6 +44,7 @@ LwF | Split MNIST | LwF.test_smnist | Stream Accuracy
 LwF | Permuted MNIST | LwF.test_pmnist | Stream Accuracy
 LwF | Split Tiny ImageNet | LwF.test_stinyimagenet | Stream Accuracy
 GSS | Split MNIST | GSS.test_smnist | Stream Accuracy
+iCaRL | iCIFAR100 | iCARL.test_iCIFAR100_batch10 | Average Incremental Accuracy
```
Comment (AndreaCossu) [inline README.md:47]: to be more aligned with avalanche, I would call the benchmark `scifar100` (so `test_scifar100`).

File: strategies/iCARL/__init__.py [hunk 2]
```diff
@@ -0,0 +1,2 @@
+from .experiment import iCARL
+#other?
```
Comment (AndreaCossu) [inline strategies/iCARL/__init__.py:2]: this is ok, you can remove the comment

File: strategies/target_results.csv [hunk 3]
```diff
@@ -14,4 +14,5 @@ gem,scifar100,0.63
 lwf,smnist,0.31
 lwf,pmnist,0.51
 lwf,stiny-imagenet,0.42
-gss,smnist,0.78
\ No newline at end of file
+gss,smnist,0.78
+iCaRL, iCIFAR100, 0.62
```
Comment (AndreaCossu) [inline strategies/target_results.csv:18]: also here, `scifar100`

File: strategies/iCARL/experiment.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
+import sys
+sys.path.append(r'/disk1/semola/avalanche')
```
Comment (AndreaCossu) [inline strategies/iCARL/experiment.py:2]: you can remove this

File: strategies/iCARL/experiment.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    """
+        Reproducing iCaRL experiments from paper
+        "iCaRL: Incremental Classifier and Representation Learning",
+        Sylvestre-Alvise Rebuffi, Alexander Kolesnikov, Georg Sperl, Christoph H. Lampert;
+        Proceedings of the IEEE Conference on
+        Computer Vision and Pattern Recognition (CVPR), 2017, pp. 2001-2010
+        https://openaccess.thecvf.com/content_cvpr_2017/html/Rebuffi_iCaRL_Incremental_Classifier_CVPR_2017_paper.html
+    """
+
+    def test_iCIFAR100_batch10(self, override_args=None):
```
Comment (AndreaCossu) [inline strategies/iCARL/experiment.py:85]: test_scifar100

File: strategies/iCARL/experiment.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+        dict_iCaRL_aia = {}
+        # Training and eval
+        for i, exp in enumerate(scenario.train_stream):
+            eval_exps = [e for e in scenario.test_stream][:i + 1]
+            strategy.train(exp, num_workers=4)
+            res = strategy.eval(eval_exps, num_workers=4)
+            dict_iCaRL_aia['Top1_Acc_Stream/Exp'+str(i)] = res['Top1_Acc_Stream/eval_phase/test_stream/Task000']
+
+        avg_ia = get_average_metric(dict_iCaRL_aia)
+        target_acc = 0.641  # get_target_result('iCaRL', 'benchmark')
```
Comment (AndreaCossu) [inline strategies/iCARL/experiment.py:187]: better to get it with the provided `get_target_result` function, to be consistent with the repository

File: strategies/iCARL/experiment.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+            fixed_memory=True, train_mb_size=config.batch_size,
+            train_epochs=config.epochs, eval_mb_size=config.batch_size,
+            plugins=[sched], device=device, evaluator=evaluator
+        )
+
+        # Dict to iCaRL Evaluation Protocol: Average Incremental Accuracy
+        dict_iCaRL_aia = {}
+        # Training and eval
+        for i, exp in enumerate(scenario.train_stream):
+            eval_exps = [e for e in scenario.test_stream][:i + 1]
```
Comment (AndreaCossu) [inline strategies/iCARL/experiment.py:181]: you can call `eval` directly on the sliced test stream `strategy.eval(scenario.test_stream[:i+1])`.

File: strategies/iCARL/experiment.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+            strategy.train(exp, num_workers=4)
+            res = strategy.eval(eval_exps, num_workers=4)
+            dict_iCaRL_aia['Top1_Acc_Stream/Exp'+str(i)] = res['Top1_Acc_Stream/eval_phase/test_stream/Task000']
+
+        avg_ia = get_average_metric(dict_iCaRL_aia)
+        target_acc = 0.641  # get_target_result('iCaRL', 'benchmark')
+        print("dict_iCaRL_aia= ", dict_iCaRL_aia)
+        print(f"iCIFAR100-batch=10 Average Incremental Accuracy: {avg_ia:.5f}")
```
Comment (AndreaCossu) [inline strategies/iCARL/experiment.py:189]: SplitCIFAR100

File: strategies/iCARL/experiment.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+
+        scenario = nc_benchmark(
+            train_dataset=train_set,
+            test_dataset=test_set,
+            n_experiences=config.nb_exp,
+            task_labels=False, seed=config.seed,
+            shuffle=False,
+            fixed_class_order=config.fixed_class_order)
+
+        evaluator = EvaluationPlugin(ExperienceAccuracy(),
```
Comment (AndreaCossu) [inline strategies/iCARL/experiment.py:156]: this is ok, but can you use `accuracy_metrics(experience=True, stream=True)`? It is the recommended way to build metrics in avalanche.

File: strategies/iCARL/experiment.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+        Proceedings of the IEEE Conference on
+        Computer Vision and Pattern Recognition (CVPR), 2017, pp. 2001-2010
+        https://openaccess.thecvf.com/content_cvpr_2017/html/Rebuffi_iCaRL_Incremental_Classifier_CVPR_2017_paper.html
+    """
+
+    def test_iCIFAR100_batch10(self, override_args=None):
+        """
+            iCIFAR-100 with 10 batches - benchmark
+        """
+        # config
```
Comment (AndreaCossu) [inline strategies/iCARL/experiment.py:89]: can you use the `create_default_args` to build the arguments? Take a look at how it is done in the other strategies (e.g., test_pmnist in EWC).

File: strategies/iCARL/experiment.py [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+                                    98, 13, 99, 7, 34, 55, 54, 26, 35, 39]
+        config.seed = 2222
+
+        #
+        device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+        torch.manual_seed(config.seed)
+        torch.cuda.manual_seed(config.seed)
+        np.random.seed(config.seed)
+        random.seed(config.seed)
+        torch.backends.cudnn.enabled = False
```
Comment (AndreaCossu) [inline strategies/iCARL/experiment.py:118]: why these cuda lines?

File: strategies/iCARL/experiment.py [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+        torch.backends.cudnn.deterministic = True
+
+        per_pixel_mean = get_dataset_per_pixel_mean(
```
Comment (AndreaCossu) [inline strategies/iCARL/experiment.py:121]: you can easily create the SplitCIFAR100 with avalanche using the `SplitCIFAR100` class directly. You can provide transformations and all the needed parameters (you can also fix the class order).

File: strategies/iCARL/experiment.py [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+            CIFAR100(expanduser("~") + "/.avalanche/data/cifar100/",
+                     train=True, download=True,
+                     transform=transforms.Compose([transforms.ToTensor()])))
+
+        transforms_group = dict(
```
Comment (AndreaCossu) [inline strategies/iCARL/experiment.py:126]: avalanche already provides transformations for cifar100. Take a look [here](https://github.com/ContinualAI/avalanche/blob/master/avalanche/benchmarks/classic/ccifar100.py), they are defined at the beginning of the file.

**Review by AndreaCossu — CHANGES_REQUESTED** [review]
Thank you @rudysemola ! I added some comments on your PR :smile:

File: README.md (hunk 1 above)
Comment (rudysemola) [reply README.md:47]: DONE

File: strategies/iCARL/__init__.py (hunk 2 above)
Comment (rudysemola) [reply strategies/iCARL/__init__.py:2]: DONE

File: strategies/target_results.csv (hunk 3 above)
Comment (rudysemola) [reply strategies/target_results.csv:18]: DONE

File: strategies/iCARL/experiment.py (hunk 4 above)
Comment (rudysemola) [reply strategies/iCARL/experiment.py:2]: DONE

File: strategies/iCARL/experiment.py (hunk 5 above)
Comment (rudysemola) [reply strategies/iCARL/experiment.py:85]: DONE

File: strategies/iCARL/experiment.py (hunk 9 above)
Comment (rudysemola) [reply strategies/iCARL/experiment.py:156]: DONE

File: strategies/iCARL/experiment.py (hunk 10 above)
Comment (rudysemola) [reply strategies/iCARL/experiment.py:89]: ok

File: strategies/iCARL/experiment.py (hunk 11 above)
Comment (rudysemola) [reply strategies/iCARL/experiment.py:118]: ok

File: strategies/iCARL/experiment.py (hunk 12 above)
Comment (rudysemola) [reply strategies/iCARL/experiment.py:97]: yep, now it is better!

File: strategies/iCARL/experiment.py (hunk 13 above)
Comment (rudysemola) [reply strategies/iCARL/experiment.py:102]: I checked, it works!

File: strategies/iCARL/experiment.py (hunk 6 above)
Comment (rudysemola) [reply strategies/iCARL/experiment.py:164]: ok

File: strategies/iCARL/experiment.py (hunk 7 above)
Comment (rudysemola) [reply strategies/iCARL/experiment.py:158]: ok if `strategy.eval(scenario.test_stream[:i+1])` is equivalent to `eval_exps = [e for e in scenario.test_stream][:i + 1]` I understand :)

File: strategies/iCARL/experiment.py (hunk 8 above)
Comment (rudysemola) [reply strategies/iCARL/experiment.py]: DONE

**Comment by rudysemola** [comment]
All the review are done :)

**Comment by AndreaCossu** [comment]
Great, thank you @rudysemola , I'll merge this :)

**Pull request merged**